### ISSUE TITLE
clicking on a desktop notification does not highlight the corresponding message

### ISSUE THREAD

**Issue by chadrik** [zulip/zulip-desktop#35]
Hi,
We're testing Zulip on Fedora 19 and so far it's working pretty well, but this issue is creating a lot of headaches.  Like the title says, if you click on a desktop notification it brings up the client window, but it does not trigger the app to display and highlight the corresponding message.  This mean that if the message lives in another stream from the currently active one you have to waste time hunting it down.

This problem does not exist with the web app using chrome desktop notifications -- the corresponding message is correctly displayed and highlighted -- however, in that case, chrome is not given proper keyboard focus, so you can't simply hit <enter> and start responding, you have to first click on it to give it focus.   Also annoying.

So I'm left trying to choose the lesser of two evils.  Ultimately, I would really like for one or the other to offer a fluid response pattern:

- click notification
- press enter to start a response in Zulip
- type message

**Comment by timabbott** [comment]
Yeah this should definitely work.  Can you open a ticket about the webapp keyboard focus issue (if there isn't one already) on the main "zulip" project issue tracker?  I suspect at least the webapp version would probably only take a few lines of code for someone to do...

**Comment by lfranchi** [comment]
I'm not sure if the alerting of the desktop app has enough hooks to do this properly :-/ Definitely something to investigate. It should be _possible_ somehow though

**Comment by chadrik** [comment]
To clarify the webapp focus issue:  it's not about keyboard focus specifically; the root problem is that the window itself is not given focus or brought to the front.   I haven't reported it because I'm not sure whether it is a general problem with chrome notifications, something specific to zulip, or something specific to linux/KDE.  I googled it, and found some Chromium bugs reported on osx that sound similar, but nothing concrete.

All that said, I'd much rather use the desktop client than the web client because the taskbar icon is invaluable for quick access.  So, any ideas on how to do this?  

Let me see if I can get the ball rolling.

It seems to me that the heart of a solution is to first associate a URL with a notification message.  To do that we have to add a URL argument to `PlatformInterface::desktopNotification` and the entire function chain leading to it.  Then we can store that value in a new member `m_lastNotificationURL`.   With that available, `PlatformInterface` can pass the URL along to the mainWindow via `trayClicked`, something like this:

```c++
    void notificationAction(uint id, QString key) {
        // Bring up the main window if the default action (usually a click on notification) was emitted.
        if (id == m_lastNotificationId && key == "default") {
            APP->mainWindow()->trayClicked(m_lastNotificationURL);
        }
    }
```

Then `trayClicked` can use the webview to set the url.  What do you think?  Would that work?

My problem is that I can't figure out how to get that url info to `ZulipWebBridge::desktopNotification`, which seems to be the start of the chain.  The chain looks like this:

```
ZulipWebBridge::desktopNotification
  (signal -> signal chain)
    HWebView::desktopNotification
      (signal -> slot)
        ZulipWindow::displayPopup
          (function call)
            PlatformInterface::desktopNotification
              (function call)
                PlatformInterface::showNotification
                  ...dbus
```

Any hints on what calls `ZulipWebBridge::desktopNotification`?   Is that coming from the Zulip server?

**Comment by excieve** [comment]
@chadrik As I understand, it essentially comes the webapp here:
https://github.com/zulip/zulip/blob/master/static/js/notifications.js#L343

So it's looks entirely possible to pass arbitrary data around.

Now, if your main concern is the window focus on Linux — it should actually work on the latest master of zulip-client. The released version doesn't have a native notifications support on Linux. See #9 and PR #25.

**Comment by excieve** [comment]
As for focusing on the notified message using the desktop client notifications, this seems to be the key:
https://github.com/zulip/zulip/blob/master/static/js/notifications.js#L319

So in this case, webapp should pass its message ID to bridge's notification function, the client should then pass it all the way through to the concrete implementation, which in turn should listen for the "click" action as pass the clicked ID back to the bridge. The bridge should then call something in the webapp that triggers the mentioned logic.

Appears possible to me but needs some cooperation between the webapp and the desktop client.

**Comment by chadrik** [comment]
Here's the line where the server should pass along the message id to the desktop client: https://github.com/zulip/zulip/blob/master/static/js/notifications.js#L343

> The bridge should then call something in the webapp that triggers the mentioned logic.

This is the missing piece.  I have no idea how to do this: I am a javascript novice and I've never worked with `QWebView`.  Anyone have a clue?

**Comment by lfranchi** [comment]
`window.bridge` is a JS object that contains method calls that end up being executed in the desktop app itself: it's a "bridge" between the JS executing in the browser and the C++ executing on the machine.

HWebView exports these functions, including `desktopNotification`, to the zulip js context so they can be triggered on demand.

**Comment by chadrik** [comment]
Hey everyone,  thanks for all the feedback.

@excieve, Sorry, I missed your first message before, hence my slightly redundant response.  The code you pointed me to made the purpose of the bridge very clear, and yup, I see that it would be quite easy to modify the bridge and server to send additional data along to the client.

I built the desktop client from master a week or two ago, and there is no issue with window focus (that was a separate chrome/webapp issue).  

My primary problem with the desktop client is the lack of message highlighting (aka narrowing), which we're discussing right now.  Where I am with that is that I know we want to somehow trigger `narrow.by_subject(message.id, {trigger: 'notification'})`, but I don't yet know how to get the `QWebView` to do that.  i.e. Is it possible for the webview to call an exported javascript function, like [received_messages](https://github.com/zulip/zulip/blob/master/static/js/notifications.js#L451)?  If no one knows the answer offhand I'll dig into the Qt docs tomorrow, but I was hoping that someone could show me the ropes.

**Comment by lfranchi** [comment]
It is possible -- in fact you can run arbitrary JS code from C++. Here is how youdo it on the mac and windows/linux:

Qt: https://github.com/zulip/zulip-desktop/blob/master/src/webview/HWebView.cpp#L111
Cocoa: https://github.com/zulip/zulip-desktop/blob/master/src/webview/HWebView_mac.mm#L194
